From 85a05fc243f1230410259d62c2a603af9efbf689 Mon Sep 17 00:00:00 2001
From: ocbuild <ocbuild@acidanthera.local>
Date: Thu, 19 May 2022 22:54:02 +0100
Subject: [PATCH] VariableDxe test patch

---
 .../Variable/RuntimeDxe/VariableDxe.c         | 52 ++++++++++++++++++-
 1 file changed, 51 insertions(+), 1 deletion(-)

diff --git a/MdeModulePkg/Universal/Variable/RuntimeDxe/VariableDxe.c b/MdeModulePkg/Universal/Variable/RuntimeDxe/VariableDxe.c
index 03fec30..c26800c 100644
--- a/MdeModulePkg/Universal/Variable/RuntimeDxe/VariableDxe.c
+++ b/MdeModulePkg/Universal/Variable/RuntimeDxe/VariableDxe.c
@@ -511,6 +511,45 @@ ProtocolIsVariablePolicyEnabled (
   return EFI_SUCCESS;
 }
 
+STATIC
+EFI_STATUS
+EFIAPI
+CreateEventMap (
+  IN       UINT32                 Type,
+  IN       EFI_TPL                NotifyTpl,
+  IN       EFI_EVENT_NOTIFY       NotifyFunction OPTIONAL,
+  IN CONST VOID                   *NotifyContext OPTIONAL,
+  IN CONST EFI_GUID               *EventGroup    OPTIONAL,
+  OUT      EFI_EVENT              *Event
+  )
+{
+  EFI_STATUS Status;
+
+  Status = EFI_UNSUPPORTED;
+  
+  if (Type == EVT_NOTIFY_SIGNAL) {
+    if (CompareGuid (EventGroup, &gEfiEventVirtualAddressChangeGuid)) {
+      Type = EVT_SIGNAL_VIRTUAL_ADDRESS_CHANGE;
+      Status = EFI_SUCCESS;
+    } else if (CompareGuid (EventGroup, &gEfiEventReadyToBootGuid)) {
+      Type = EVT_SIGNAL_EXIT_BOOT_SERVICES;
+      Status = EFI_SUCCESS;
+    }
+  }
+
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  return gBS->CreateEvent (
+    Type,
+    NotifyTpl,
+    NotifyFunction,
+    (VOID *)NotifyContext,
+    Event
+  );
+}
+
 /**
   Variable Driver main entry point. The Variable driver places the 4 EFI
   runtime services in the EFI System Table and installs arch protocols
@@ -532,7 +571,8 @@ VariableServiceInitialize (
 {
   EFI_STATUS  Status;
   EFI_EVENT   ReadyToBootEvent;
-  EFI_EVENT   EndOfDxeEvent;
+  VOID        *OriginalCreateEventEx;
+  UINT32      Revision;
 
   Status = VariableCommonInitialize ();
   ASSERT_EFI_ERROR (Status);
@@ -587,6 +627,9 @@ VariableServiceInitialize (
     VariableWriteServiceInitializeDxe ();
   }
 
+  OriginalCreateEventEx = gBS->CreateEventEx;
+  gBS->CreateEventEx = CreateEventMap;
+
   Status = gBS->CreateEventEx (
                   EVT_NOTIFY_SIGNAL,
                   TPL_NOTIFY,
@@ -600,6 +643,8 @@ VariableServiceInitialize (
   //
   // Register the event handling function to reclaim variable for OS usage.
   //
+  Revision = gST->Hdr.Revision;
+  gST->Hdr.Revision = EFI_2_00_SYSTEM_TABLE_REVISION;
   Status = EfiCreateEventReadyToBootEx (
              TPL_NOTIFY,
              OnReadyToBoot,
@@ -607,10 +652,12 @@ VariableServiceInitialize (
              &ReadyToBootEvent
              );
   ASSERT_EFI_ERROR (Status);
+  gST->Hdr.Revision = Revision;
 
   //
   // Register the event handling function to set the End Of DXE flag.
   //
+#if 0
   Status = gBS->CreateEventEx (
                   EVT_NOTIFY_SIGNAL,
                   TPL_CALLBACK,
@@ -620,6 +667,7 @@ VariableServiceInitialize (
                   &EndOfDxeEvent
                   );
   ASSERT_EFI_ERROR (Status);
+#endif
 
   // Register and initialize the VariablePolicy engine.
   Status = InitVariablePolicyLib (VariableServiceGetVariable);
@@ -634,5 +682,7 @@ VariableServiceInitialize (
                   );
   ASSERT_EFI_ERROR (Status);
 
+  gBS->CreateEventEx = OriginalCreateEventEx;
+
   return EFI_SUCCESS;
 }
-- 
2.31.1

